crypto3/libs/zk/include/nil/crypto3/zk/snark/systems/plonk/placeholder/lookup_argument.hpp

```diff
@@ -173,268 +119,138 @@ namespace nil {
-                        F_dfs[1] = preprocessed_data.q_last * ( V_L * V_L - V_L );
-
-                        // Polynomial g is waaay too large, saving memory here, by making code very unreadable.
-                        //F_dfs[2] = (one_polynomial - (preprocessed_data.q_last + preprocessed_data.q_blind)) *
-                        //           (V_L_shifted * h - V_L * g);
-                        F_dfs[2] = zero_polynomial;
-                        if( part_sizes.size() == 1 ){
-                            auto &g = gs[0];
-                            auto &h = hs[0];
-                            g *= V_L;
-                            h *= V_L_shifted;
-                            g -= h;
-                            h = polynomial_dfs_type(); // just clean the memory of h.
-                            g *= (preprocessed_data.q_last + preprocessed_data.q_blind) - one_polynomial;
-                            F_dfs[2] = std::move(g);
-                        } else {
-                            std::vector<polynomial_dfs_type> parts;
-                            BOOST_ASSERT(part_sizes.size() == gs.size());
-                            BOOST_ASSERT(part_sizes.size() == hs.size());
-                            BOOST_ASSERT(part_sizes.size() == lookup_alphas.size() + 1);
-
-                            polynomial_dfs_type current_poly = V_L;
-                            polynomial_dfs_type previous_poly = V_L;
-
-                            for( std::size_t i = 0; i < lookup_alphas.size(); i ++ ){
-                                auto &g = gs[i];
-                                auto &h = hs[i];
-                                auto reduced_g = reduce_dfs_polynomial_domain(g, basic_domain->m);
-                                auto reduced_h = reduce_dfs_polynomial_domain(h, basic_domain->m);
-                                for( std::size_t j = 0; j < assignment_desc.usable_rows_amount; j++){
-                                    current_poly[j] = (previous_poly[j] * reduced_g[j]) * reduced_h[j].inversed();
-                                }
-                                commitment_scheme.append_to_batch(PERMUTATION_BATCH, current_poly);
-                                auto par = lookup_alphas[i] * (previous_poly * g - current_poly * h);
-                                F_dfs[2] += par;
-                                previous_poly = current_poly;
-                            }
-                            std::size_t last = lookup_alphas.size();
-                            auto &g = gs[last];
-                            auto &h = hs[last];
-                            F_dfs[2] += (previous_poly * g - V_L_shifted * h);
-                            F_dfs[2] *= (preprocessed_data.q_last + preprocessed_data.q_blind) - one_polynomial;
+                        // Create a constraint for H_i(X) * (alpha - F_i(X)) + 1 == 0.
+                        std::vector<typename FieldType::value_type> h_challenges =
```

Comment:
I think, we can use only one challenge's degrees, that would be more effective for verification.

Reply:
Changed.